File: benchmarks/typescript/benchmark.ts
import { performance } from "perf_hooks";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function emit(name: string, timeMs: number, result: string): void {
  const json = JSON.stringify({
    language: "typescript",
    benchmark: name,
    time_ms: parseFloat(timeMs.toFixed(4)),
    result,
  });
  console.log(json);
}

// ---------------------------------------------------------------------------
// 1. fib  --  naive recursive Fibonacci(40)
// ---------------------------------------------------------------------------

function fib(n: number): number {
  if (n <= 1) return n;
  return fib(n - 1) + fib(n - 2);
}

function benchFib(): void {
  const start = performance.now();
  const result = fib(40);
  const elapsed = performance.now() - start;
  emit("fib", elapsed, String(result));
}

// ---------------------------------------------------------------------------
// 2. trees  --  binary tree depth 21, allocate + checksum
// ---------------------------------------------------------------------------

interface TreeNode {
  left: TreeNode | null;
  right: TreeNode | null;
}

function makeTree(depth: number): TreeNode {
  if (depth === 0) return { left: null, right: null };
  return { left: makeTree(depth - 1), right: makeTree(depth - 1) };
}

function checksum(node: TreeNode | null): number {
  if (node === null) return 0;
  if (node.left === null) return 1;
  return 1 + checksum(node.left) + checksum(node.right);
}

function benchTrees(): void {
  const depth = 21;
  const start = performance.now();
  const tree = makeTree(depth);
  const cs = checksum(tree);
  const elapsed = performance.now() - start;
  emit("trees", elapsed, String(cs));
}

// ---------------------------------------------------------------------------
// 3. matrix  --  1000x1000 matrix multiply, result = C[0][0]
// ---------------------------------------------------------------------------

function benchMatrix(): void {
  const N = 1000;

  // Allocate and fill A and B with deterministic values
  const A: number[][] = new Array(N);
  const B: number[][] = new Array(N);
  for (let i = 0; i < N; i++) {
    A[i] = new Array(N);
    B[i] = new Array(N);
    for (let j = 0; j < N; j++) {
      A[i][j] = (i * N + j) % 100;
      B[i][j] = (j * N + i) % 100;
    }
  }

  const start = performance.now();

  // C = A * B
  const C: number[][] = new Array(N);
  for (let i = 0; i < N; i++) {
    C[i] = new Array(N).fill(0);
    for (let k = 0; k < N; k++) {
      const aik = A[i][k];
      for (let j = 0; j < N; j++) {
        C[i][j] += aik * B[k][j];
      }
    }
  }

  const elapsed = performance.now() - start;
  emit("matrix", elapsed, String(C[0][0]));
}

// ---------------------------------------------------------------------------
// 4. strings  --  1 MB deterministic ASCII: word count, count "the", reverse,
//                 simple hash.  Result = word count.
// ---------------------------------------------------------------------------

function benchStrings(): void {
  // Build a deterministic ~1 MB string from a fixed vocabulary
  const words = [
    "the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
    "and", "the", "cat", "sat", "on", "the", "mat", "while", "birds", "fly",
    "high", "above", "the", "clouds", "in", "the", "bright", "blue", "sky",
  ];

  const parts: string[] = [];
  let totalLen = 0;
  const target = 1_000_000;
  let idx = 0;
  while (totalLen < target) {
    const w = words[idx % words.length];
    if (totalLen > 0) {
      parts.push(" ");
      totalLen += 1;
    }
    parts.push(w);
    totalLen += w.length;
    idx++;
  }
  const text = parts.join("");

  const start = performance.now();

  // Word count
  const wordList = text.split(/\s+/);
  const wordCount = wordList.length;

  // Count occurrences of "the"
  let theCount = 0;
  for (const w of wordList) {
    if (w === "the") theCount++;
  }

  // Reverse the whole string
  const reversed = text.split("").reverse().join("");

  // Simple hash (djb2-style)
  let hash = 5381;
  for (let i = 0; i < reversed.length; i++) {
    hash = ((hash * 33) ^ reversed.charCodeAt(i)) >>> 0;
  }

  // Touch theCount and hash so they are not optimised away
  void theCount;
  void hash;

  const elapsed = performance.now() - start;
  emit("strings", elapsed, String(wordCount));
}

// ---------------------------------------------------------------------------
// 5. concurrent  --  Promise.all with 1000 promises each computing fib(30).
//    JS is single-threaded, so they run sequentially.  Result = sum.
// ---------------------------------------------------------------------------

function fibSmall(n: number): number {
  if (n <= 1) return n;
  return fibSmall(n - 1) + fibSmall(n - 2);
}

async function benchConcurrent(): Promise<void> {
  const N = 1000;
  const start = performance.now();

  const promises: Promise<number>[] = [];
  for (let i = 0; i < N; i++) {
    promises.push(Promise.resolve().then(() => fibSmall(30)));
  }

  const results = await Promise.all(promises);
  let sum = 0;
  for (const r of results) sum += r;

  const elapsed = performance.now() - start;
  emit("concurrent", elapsed, String(sum));
}

// ---------------------------------------------------------------------------
// CLI dispatcher
// ---------------------------------------------------------------------------

const benchmarks: Record<string, () => void | Promise<void>> = {
  fib: benchFib,
  trees: benchTrees,
  matrix: benchMatrix,
  strings: benchStrings,
  concurrent: benchConcurrent,
};

async function main(): Promise<void> {
  const arg = process.argv[2] ?? "all";

  if (arg === "all") {
    for (const name of Object.keys(benchmarks)) {
      await benchmarks[name]();
    }
  } else if (benchmarks[arg]) {
    await benchmarks[arg]();
  } else {
    console.error(`Unknown benchmark: ${arg}`);
    console.error(`Available: ${Object.keys(benchmarks).join(", ")}, all`);
    process.exit(1);
  }
}

main();

File: benchmarks/csharp/Benchmark.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace Benchmarks;

// ---------------------------------------------------------------------------
// JSON helper
// ---------------------------------------------------------------------------
static class JsonResult
{
    public static void Print(string benchmark, double timeMs, string result)
    {
        string time = timeMs.ToString("F4", CultureInfo.InvariantCulture);
        Console.WriteLine(
            $"{{\"language\":\"csharp\",\"benchmark\":\"{benchmark}\",\"time_ms\":{time},\"result\":\"{result}\"}}");
    }
}

// ---------------------------------------------------------------------------
// 1. fib  --  naive recursive fibonacci(40)
// ---------------------------------------------------------------------------
static class FibBenchmark
{
    static long Fib(int n)
    {
        if (n < 2) return n;
        return Fib(n - 1) + Fib(n - 2);
    }

    public static void Run()
    {
        var sw = Stopwatch.StartNew();
        long result = Fib(40);
        sw.Stop();
        JsonResult.Print("fib", sw.Elapsed.TotalMilliseconds, result.ToString());
    }
}

// ---------------------------------------------------------------------------
// 2. trees  --  binary tree of depth 21, class-based nodes, checksum
// ---------------------------------------------------------------------------
static class TreesBenchmark
{
    sealed class TreeNode
    {
        public TreeNode? Left;
        public TreeNode? Right;

        public TreeNode(TreeNode? left, TreeNode? right)
        {
            Left = left;
            Right = right;
        }
    }

    static TreeNode BuildTree(int depth)
    {
        if (depth == 0) return new TreeNode(null, null);
        return new TreeNode(BuildTree(depth - 1), BuildTree(depth - 1));
    }

    static long Checksum(TreeNode node)
    {
        if (node.Left == null) return 1;
        return 1 + Checksum(node.Left) + Checksum(node.Right!);
    }

    public static void Run()
    {
        var sw = Stopwatch.StartNew();
        var tree = BuildTree(21);
        long cs = Checksum(tree);
        sw.Stop();
        JsonResult.Print("trees", sw.Elapsed.TotalMilliseconds, cs.ToString());
    }
}

// ---------------------------------------------------------------------------
// 3. matrix  --  1000x1000 double matrix multiply  C = A * B
// ---------------------------------------------------------------------------
static class MatrixBenchmark
{
    public static void Run()
    {
        const int N = 1000;

        // Deterministic fill
        double[][] a = new double[N][];
        double[][] b = new double[N][];
        double[][] c = new double[N][];

        for (int i = 0; i < N; i++)
        {
            a[i] = new double[N];
            b[i] = new double[N];
            c[i] = new double[N];
            for (int j = 0; j < N; j++)
            {
                a[i][j] = (i * N + j) * 1.0e-6;
                b[i][j] = (j * N + i) * 1.0e-6;
            }
        }

        var sw = Stopwatch.StartNew();

        // Classic ijk multiply
        for (int i = 0; i < N; i++)
        {
            for (int k = 0; k < N; k++)
            {
                double aik = a[i][k];
                for (int j = 0; j < N; j++)
                {
                    c[i][j] += aik * b[k][j];
                }
            }
        }

        sw.Stop();

        string result = c[0][0].ToString("G17", CultureInfo.InvariantCulture);
        JsonResult.Print("matrix", sw.Elapsed.TotalMilliseconds, result);
    }
}

// ---------------------------------------------------------------------------
// 4. strings  --  1 MB deterministic ASCII text processing
// ---------------------------------------------------------------------------
static class StringsBenchmark
{
    public static void Run()
    {
        // Build 1 MB of deterministic ASCII text
        string[] words = {
            "the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
            "alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta", "theta",
            "iota", "kappa", "lambda", "mu", "nu", "xi", "omicron", "pi", "rho",
            "sigma", "tau", "upsilon", "phi", "chi", "psi", "omega"
        };

        const int targetSize = 1024 * 1024; // 1 MB
        var sb = new StringBuilder(targetSize + 256);
        int idx = 0;
        while (sb.Length < targetSize)
        {
            if (sb.Length > 0) sb.Append(' ');
            sb.Append(words[idx % words.Length]);
            idx++;
        }
        string text = sb.ToString();

        var sw = Stopwatch.StartNew();

        // Count words
        int wordCount = 0;
        bool inWord = false;
        for (int i = 0; i < text.Length; i++)
        {
            if (text[i] == ' ')
            {
                inWord = false;
            }
            else if (!inWord)
            {
                inWord = true;
                wordCount++;
            }
        }

        // Count occurrences of "the"
        int theCount = 0;
        int searchFrom = 0;
        while (true)
        {
            int pos = text.IndexOf("the", searchFrom, StringComparison.Ordinal);
            if (pos < 0) break;
            // Only count whole words
            bool leftOk = pos == 0 || text[pos - 1] == ' ';
            bool rightOk = pos + 3 >= text.Length || text[pos + 3] == ' ';
            if (leftOk && rightOk) theCount++;
            searchFrom = pos + 1;
        }

        // Reverse the string
        char[] chars = text.ToCharArray();
        Array.Reverse(chars);
        string reversed = new string(chars);

        // SHA-256 hash of the reversed string
        byte[] hashBytes = SHA256.HashData(Encoding.UTF8.GetBytes(reversed));
        StringBuilder hashSb = new StringBuilder(64);
        foreach (byte b in hashBytes)
            hashSb.Append(b.ToString("x2"));
        string hash = hashSb.ToString();

        sw.Stop();

        // Prevent dead-code elimination: use theCount, reversed.Length, hash
        _ = theCount;
        _ = reversed.Length;
        _ = hash;

        JsonResult.Print("strings", sw.Elapsed.TotalMilliseconds, wordCount.ToString());
    }
}

// ---------------------------------------------------------------------------
// 5. concurrent  --  Task.WhenAll with 1000 tasks computing fib(30)
// ---------------------------------------------------------------------------
static class ConcurrentBenchmark
{
    static long Fib(int n)
    {
        if (n < 2) return n;
        return Fib(n - 1) + Fib(n - 2);
    }

    public static void Run()
    {
        const int taskCount = 1000;

        var sw = Stopwatch.StartNew();

        var tasks = new Task<long>[taskCount];
        for (int i = 0; i < taskCount; i++)
        {
            tasks[i] = Task.Run(() => Fib(30));
        }
        Task.WhenAll(tasks).GetAwaiter().GetResult();

        long sum = 0;
        for (int i = 0; i < taskCount; i++)
        {
            sum += tasks[i].Result;
        }

        sw.Stop();
        JsonResult.Print("concurrent", sw.Elapsed.TotalMilliseconds, sum.ToString());
    }
}

// ---------------------------------------------------------------------------
// CLI dispatcher
// ---------------------------------------------------------------------------
class Program
{
    static readonly Dictionary<string, Action> Benchmarks = new()
    {
        ["fib"]        = FibBenchmark.Run,
        ["trees"]      = TreesBenchmark.Run,
        ["matrix"]     = MatrixBenchmark.Run,
        ["strings"]    = StringsBenchmark.Run,
        ["concurrent"] = ConcurrentBenchmark.Run,
    };

    static void Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.Error.WriteLine("Usage: Benchmark <name|all> [name ...]");
            Console.Error.WriteLine("Available: " + string.Join(", ", Benchmarks.Keys));
            Environment.Exit(1);
        }

        var toRun = new List<string>();

        foreach (string arg in args)
        {
            if (arg.Equals("all", StringComparison.OrdinalIgnoreCase))
            {
                toRun.AddRange(Benchmarks.Keys);
            }
            else if (Benchmarks.ContainsKey(arg))
            {
                toRun.Add(arg);
            }
            else
            {
                Console.Error.WriteLine($"Unknown benchmark: {arg}");
                Console.Error.WriteLine("Available: " + string.Join(", ", Benchmarks.Keys));
                Environment.Exit(1);
            }
        }

        foreach (string name in toRun)
        {
            Benchmarks[name]();
        }
    }
}
